File: ios/RecordStore.swift
//
//  SRecordStore.swift
//  openland
//
//  Created by Steve Kite on 5/10/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation

class ChangedRecord {
  let key: String
  let fields: Set<String>
  init(key: String, fields: Set<String>) {
    self.key = key
    self.fields = fields
  }
}

class RecordStore {
  private var inMemory: [String: Record] = [:]
  
  //
  // Loading
  //
  
  func isInMemory(key: String) -> Bool {
    return inMemory[key] != nil
  }
  
  func loaded(recordSet: RecordSet) {
    for r in recordSet {
      loaded(record: r.value)
    }
  }
  
  func loaded(record: Record) {
    if self.inMemory[record.key] != nil {
      fatalError("Record " + record.key + " already loaded")
    }
    self.inMemory[record.key] = record
  }
  
  //
  // Read
  //
  
  func read(key: String) -> Record {
    let res = self.inMemory[key]
    if res == nil {
      let rs = Record(key: key, fields: [:])
      self.inMemory[key] = rs
      return rs
      // fatalError("Record " + key + " is not loaded yet")
    } else {
      return res!
    }
  }
  
  //
  // Merging
  //
  
  func merge(recordSet: RecordSet) -> [String: ChangedRecord] {
    let res = SharedDictionary<String, ChangedRecord>()
    for k in recordSet {
      merge(record: k.value, changed: res)
    }
    return res.dict
  }
  
  func merge(record: Record) -> [String: ChangedRecord] {
    let res = SharedDictionary<String, ChangedRecord>()
    merge(record: record, changed: res)
    return res.dict
  }
  
  private func merge(record: Record, changed: SharedDictionary<String, ChangedRecord>) {
    var fields: [String: RecordValue] = [:]
    var changedFields: Set<String> = Set()
    let ex = inMemory[record.key]
    if ex != nil {
      for f in ex!.fields {
        fields[f.key] = f.value
      }
    }
    for f in record.fields {
      let exf = ex?.fields[f.key]
      if exf == nil || exf != f.value {
        changedFields.insert(f.key)
        fields[f.key] = f.value
      }
    }
    if changedFields.count > 0 {
      inMemory[record.key] = Record(key: record.key, fields: fields)
      changed[record.key] = ChangedRecord(key: record.key, fields: changedFields)
    }
  }
}

File: packages/openland-graphql/spacex/store/RecordStore.ts

export type RecordValueString = {
    type: 'string',
    value: string
};
export type RecordValueNumber = {
    type: 'number',
    value: number
};
export type RecordValueBoolean = {
    type: 'boolean',
    value: boolean
};
export type RecordValueNull = {
    type: 'null'
};
export type RecordValueReference = {
    type: 'reference',
    key: string
};

export type RecordValueList = {
    type: 'list',
    values: RecordValue[]
};

export type RecordValue =
    | RecordValueString
    | RecordValueNumber
    | RecordValueBoolean
    | RecordValueNull
    | RecordValueReference
    | RecordValueList;

export type Record = {
    key: string;
    fields: { [key: string]: RecordValue }
};

export type RecordSet = { [key: string]: Record };

export type ChangedRecord = { [key: string]: string };

function isRecordValueEquals(a: RecordValue, b: RecordValue) {
    if (a.type !== b.type) {
        return false;
    }

    if (a.type === 'number' && b.type === 'number') {
        return a.value === b.value;
    }
    if (a.type === 'boolean' && b.type === 'boolean') {
        return a.value === b.value;
    }
    if (a.type === 'string' && b.type === 'string') {
        return a.value === b.value;
    }
    if (a.type === 'null' && b.type === 'null') {
        return true;
    }
    if (a.type === 'reference' && b.type === 'reference') {
        return a.key === b.key;
    }
    if (a.type === 'list' && b.type === 'list') {
        let la = a.values;
        let lb = b.values;
        if (la.length !== lb.length) {
            return false;
        }
        for (let i = 0; i < la.length; i++) {
            if (!isRecordValueEquals(la[i], lb[i])) {
                return false;
            }
        }
        return true;
    }

    return true;
}

export class RecordStore {
    private inMemory: { [key: string]: Record } = {};

    //
    // Loading
    //

    isInMemory = (key: string) => {
        return !!this.inMemory[key];
    }

    loaded = (set: RecordSet) => {
        for (let k in set) {
            if (set.hasOwnProperty(k)) {
                this.loadedRecord(set[k]);
            }
        }
    }

    loadedRecord = (record: Record) => {
        if (!!this.inMemory[record.key]) {
            throw Error('Record ' + record.key + ' already loaded');
        }
        this.inMemory[record.key] = record;
    }

    //
    // Read
    //

    read = (key: string): Record => {
        if (!this.inMemory[key]) {
            throw Error('Record ' + key + ' not loaded yet');
        }
        return this.inMemory[key]!;
    }

    //
    // Merging
    //

    merge = (set: RecordSet): { [key: string]: ChangedRecord } => {
        let res: { [key: string]: ChangedRecord } = {};
        for (let k in set) {
            if (set.hasOwnProperty(k)) {
                this._merge(set[k], res);
            }
        }
        return res;
    }

    mergeRecord = (record: Record): { [key: string]: ChangedRecord } => {
        let res: { [key: string]: ChangedRecord } = {};
        this._merge(record, res);
        return res;
    }

    private _merge = (record: Record, changed: { [key: string]: ChangedRecord }) => {
        let fields: { [key: string]: RecordValue } = {};
        let changedFields: { [key: string]: string } = {};
        let ex = this.inMemory[record.key];
        if (ex) {
            for (let f in ex.fields) {
                if (ex.fields.hasOwnProperty(f)) {
                    fields[f] = ex.fields[f];
                }
            }
        }

        for (let f in record.fields) {
            if (record.fields.hasOwnProperty(f)) {
                let exf = ex ? ex.fields[f] : null;
                if (!exf || !isRecordValueEquals(exf, record.fields[f])) {
                    changedFields[f] = f;
                    fields[f] = record.fields[f];
                }
            }
        }

        if (Object.keys(changedFields).length > 0) {
            this.inMemory[record.key] = { key: record.key, fields };
            changed[record.key] = changedFields;
        }
    }
}